jpa/persistence-cdi-extension/src/main/java/org/jboss/as/jpa/beanmanager/IntegratePersistenceAfterBeanDiscovery.java

```diff
@@ -0,0 +1,444 @@
+        }
+        beanConfigurator.beanClass(PersistenceUnitUtil.class);
+        beanConfigurator.produceWith(c -> {
+                    return integrationWithCDIBag.getEntityManagerFactory().getPersistenceUnitUtil();
+                }
+        );
+
+    }
+
+    private void cache(
+            AfterBeanDiscovery afterBeanDiscovery,
+            PersistenceUnitMetadata persistenceUnitMetadata,
+            List<String> qualifiers,
+            IntegrationWithCDIBag integrationWithCDIBag) {
+
+        String scope = dependentScoped;
+
+        BeanConfigurator<Cache> beanConfigurator = afterBeanDiscovery.addBean();
+        beanConfigurator.addTransitiveTypeClosure(Cache.class);
+
+        Class<? extends Annotation> scopeAnnotation = null;
+        try {
+            scopeAnnotation = persistenceUnitMetadata.getClassLoader().loadClass(scope).asSubclass(Annotation.class);
+        } catch (ClassNotFoundException e) {
+            throw JpaLogger.ROOT_LOGGER.classNotFound(e, "Cache", persistenceUnitMetadata.getPersistenceUnitName(), persistenceUnitMetadata.getScopedPersistenceUnitName());
+        }
+        beanConfigurator.scope(scopeAnnotation);
+
+        for (String qualifier : qualifiers) {
+            final Class<? extends Annotation> qualifierType;
+            try {
+                qualifierType = persistenceUnitMetadata.getClassLoader()
+                        .loadClass(qualifier)
+                        .asSubclass(Annotation.class);
+            } catch (ClassNotFoundException e) {
+                throw JpaLogger.ROOT_LOGGER.classNotFound(e, "Cache", persistenceUnitMetadata.getPersistenceUnitName(), persistenceUnitMetadata.getScopedPersistenceUnitName());
+            }
+            beanConfigurator.addQualifier(ScopeProxy.createProxy(qualifierType));
+        }
+        beanConfigurator.beanClass(Cache.class);
+        beanConfigurator.produceWith(c -> {
+                    return integrationWithCDIBag.getEntityManagerFactory().getCache();
+                }
+        );
+
+    }
+
+
+    private void metamodel(
+            AfterBeanDiscovery afterBeanDiscovery,
+            PersistenceUnitMetadata persistenceUnitMetadata,
+            List<String> qualifiers,
+            IntegrationWithCDIBag integrationWithCDIBag) {
+
+        String scope = dependentScoped;
+
+        BeanConfigurator<Metamodel> beanConfigurator = afterBeanDiscovery.addBean();
+        beanConfigurator.addTransitiveTypeClosure(Metamodel.class);
+
+        Class<? extends Annotation> scopeAnnotation = null;
+        try {
+            scopeAnnotation = persistenceUnitMetadata.getClassLoader().loadClass(scope).asSubclass(Annotation.class);
+        } catch (ClassNotFoundException e) {
+            throw JpaLogger.ROOT_LOGGER.classNotFound(e, "Metamodel", persistenceUnitMetadata.getPersistenceUnitName(), persistenceUnitMetadata.getScopedPersistenceUnitName());
+        }
+        beanConfigurator.scope(scopeAnnotation);
```

Comment:
This could simply be `beanConfiguator.scope(Dependent.class)`. This would save the class loading as the scope from this producer cannot be changed.

Reply:
I'll do the same for the extra loading of the ApplicationScoped class as well.